Add Helm chart

stevehipwell opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This PR adds an official _Helm_ chart based on the previous [stable chart](https://github.com/helm/charts/tree/master/stable/metrics-server) and the deployment yaml from this project.

This chart has been given a major version bump and should be updatable from the existing stable chart but without support for additional containers and volumes (this could be re-added but seemed un-needed).

**What is outstanding**:

The following items need some discussion.

- CI/CD automation
- Chart maintainer

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #572.

pierluigilenoci commented:
@s-urbaniak @brancz @serathius could you please take a look?

stevehipwell commented:
Has anyone had a chance to look at this yet?

pierluigilenoci commented:
@s-urbaniak @brancz @serathius please! ❤️

pierluigilenoci commented:
@stevehipwell this PR is still in draft, maybe makes sense to try to promote it as real PR to be considered?

stevehipwell commented:
@pierluigilenoci the actual chart is complete but I've got outstanding questions around the CI/CD tool to use (I'm familiar with GitHub actiuons for this) and who the chart maintainer needs to be set as. I'm happy to write the readme now and make it an official PR but I was waiting for answers.

pgollucci commented:
commenting to track

stevehipwell commented:
I've added the chart README and setup GitHub actions for the automation.

The chart lint action will run on PRs when the `charts/**` path has been modified. The chart will be released when there is a release published with a `v` prefix. The chart will be released on a tag following the pattern `metrics-server-helm-chart-${VERSION}` where `VERSION` is the chart version. Chart version and app version will both need to be incremented as part of the release process.

krmichelos commented:
> The chart will be released when there is a release published with a `v` prefix. The chart will be released on a tag following the pattern `metrics-server-helm-chart-${VERSION}`

FYI, I have seen in a number of projects where the chart releaser action is triggered when any file in the charts directory is changed on master and will just fail until the chart version in the Chart.yaml is updated to a version that hasn't already been released

stevehipwell commented:
@krmichelos that's not a pattern I'm familiar with, specifically the failing part. If a repo is just for _Helm_ charts then releases tend to happen on any change to master where the chart version has been changed (this works with multiple charts). For a repo like this there needs to be a specific trigger and the one I've configured would seem to fit in best with the project release process (for the descheduler I used a different pattern as they have a different release process).

stevehipwell commented:
I've rebased, is there anything else I need to do to get this progressed?

stevehipwell commented:
@brancz @serathius could you take a look?

pierluigilenoci commented:
@brancz @serathius could you take a look?

mohsen0 commented:
@brancz @serathius could this be looked at?
Thanks

serathius commented:
Hey,
I discussed maintenance of Helm charts with @brancz on recent SIG Instrumentation triage meetings. Helm chart was already introduced for kube-state-metrics, so we would want to have consistent approach to Helm throughout SIG Instrumentation projects.

One of the requirements put was that Helm chart in this repo should not affect maintenance of non-helm parts. Currently there is a standing issue of spamy helm releases https://github.com/kubernetes/kube-state-metrics/issues/1392 that is considered breaking aforementioned requirement https://github.com/kubernetes/kube-state-metrics/issues/1392#issuecomment-784941016

As I don't have knowledge about Helm nor GitHub actions, please let me know if this PR would introduce similar issue as https://github.com/kubernetes/kube-state-metrics/issues/1392 if so I would consider solving this issue as a blocker.

stevehipwell commented:
@serathius I'm not sure that I agree with [kubernetes/kube-state-metrics #1392](https://github.com/kubernetes/kube-state-metrics/issues/1392) that _Helm_ chart releases are spam. But if this is the view taken then based on the current state of GitHub notifications you'd need to move the _Helm_ charts into another repo if you don't want any notifications or use a release window to limit changes.

The workflow designed for this _Helm_ chart should make sure that the chart releases (`metrics-server-helm-chart-1.0.0` are easily distinguished from the actual releases (`v1.0.0`). For a project such as this where the _Helm_ chart is based on the repo content (the deployment yaml) I wouldn't expect much if any changes between core releases.

Personally I'd take this being merged in with only a single chart release per core release (assuming breaking changes would be fixed) as better than the current situation where no chart changes can be made.

pierluigilenoci commented:
@serathius is there any hope of finally seeing the metrics-server chart released?

SM616 commented:
Is there any progress on this PR?

stevehipwell commented:
@SM616 I'm still waiting for a review.

stevehipwell commented:
I've released my own version of the chart which I will deprecate when this is merged.

https://artifacthub.io/packages/helm/stevehipwell/metrics-server
https://github.com/stevehipwell/helm-charts/tree/master/charts/metrics-server

pierluigilenoci commented:
@serathius @brancz could you please reviews this and maybe approve it?

elcomtik commented:
@stevehipwell I think that maintainers don't want to have a helm chart in this repo. They mentioned that they prefer the usage of kustomized in some related issues.

I would like to have the official helm chart also, but I think it will need to be a separate repository.

stevehipwell commented:
@elcomtik see https://github.com/kubernetes-sigs/metrics-server/pull/670#issuecomment-816590042.

elcomtik commented:
> @elcomtik see [#670 (comment)](https://github.com/kubernetes-sigs/metrics-server/pull/670#issuecomment-816590042).

I'm aware of it, however, it cannot be considered official. It may be better quality compared to bitnami chart, but it won't be suitable for some corporate policies mentioned above.

@brancz @serathius What about the option to move your repo under kubernetes-sigs organization namespace? Did someone consider the scenario of the separate repo?

stevehipwell commented:
@elcomtik a _helm-charts_ repo under _kubernetes-sigs_ would be the next best solution to having the carts live with their specific components. I'd be happy to open a PR to configure such a repo if it was created.

serathius commented:
Organization _kubernetes-sigs_ was created to collect subprojects from Kubernetes Special Interest Groups (SIGs). I think you should bring this topic to [SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps).

stevehipwell commented:
@serathius I've opened an issue https://github.com/kubernetes/community/issues/5712.

stevehipwell commented:
@serathius @brancz there doesn't seem to be any traction for a central Helm chart repo so this looks like the best fit for the community. This chart needs minimal (if any) changes besides the version and other changes provided in the yaml, as can be seen by when I've needed to update the PR. Furthermore as it stands the chart will only be published when a new version of Metrics Server is published on a `v` tag so there will not be any additional "noise" around releases.

stevehipwell commented on charts/metrics-server/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+name: metrics-server
+description: Metrics Server is a scalable, efficient source of container resource metrics for Kubernetes built-in autoscaling pipelines.
+type: application
+version: 3.0.0
```
Originally I was keeping this separate, but that's a good point. I can make that change if you'd like?

pierluigilenoci replied:
https://media.giphy.com/media/GcSqyYa2aF8dy/giphy.gif

KlavsKlavsen commented:
Here's to hoping this gets resolved soon.. still running chart v2.11.4 - as I'd otherwise have to switch to one of the forks of this chart :(

stevehipwell commented:
@KlavsKlavsen this chart is available at https://artifacthub.io/packages/helm/stevehipwell/metrics-server from my repo until this is resolved.

pierluigilenoci commented:
@brancz @serathius please take a look.

elcomtik commented:
@stevehipwell @pierluigilenoci I'm curious why do you keep knocking on closed doors. They obviously aren't interested in this contribution.

jbuettnerbild commented:
that is sad because i would be very happy on an official helm chart

stevehipwell commented:
@elcomtik there is a problem that needs solving, people want to use Helm and they want the official image and pattern. Sig apps say the chart should be in this repo, which would appear to be the general rule for the other repos, so this is open and updated.

petewilcock commented:
@stevehipwell +1 - I can't see the aversion to an official chart for something so commonly used. Your willingness to maintain this is to your credit and many developers would love to have someone so independently motivated to add utility to their app!

elcomtik commented:
> @elcomtik there is a problem that needs solving, people want to use Helm and they want the official image and pattern. Sig apps say the chart should be in this repo, which would appear to be the general rule for the other repos, so this is open and updated.

@stevehipwell I agree with you, that this problem needs to be solved. I'm not against you and really appreciate your determination. There is an issue that maintainers of this repo seem to not respect the will of SIG apps in spite of fact that they sent you to ask them.

pierluigilenoci commented:
> @stevehipwell I agree with you, that this problem needs to be solved. I'm not against you and really appreciate your determination. There is an issue that maintainers of this repo seem to not respect the will of SIG apps in spite of fact that they sent you to ask them.

@elcomtik finding a place to save the software chart is really trivial stuff, it's a senseless obstruction.

This repo is not a personal repo of the owners but it is a community repo. If the current owners no longer agree with the rest of the SIG community, they can argue about that or you eventually step aside. Doing stonewalling is not a mature attitude.

@stevehipwell proposed multiple solutions but he only gets doors slammed in the face. Don't be surprised if people stop contributing.

Obviously, I never lose gratitude towards those who are committed to the community.

pierluigilenoci commented:
/assign s-urbaniak

pierluigilenoci commented:
/assigned dgrisonnet

pierluigilenoci commented:
/assign dgrisonnet

pierluigilenoci commented:
Now all the owners are involved, let's see what comes out. 😛

pierluigilenoci commented:
If you tell the whole story it becomes like a joke: the metrics-server helm chart needs a new home since November 13th 2019, when the helm chart repo was deprecated... 555 days to find a place and counting. 🤦🏻

serathius commented:
I think the original point from https://github.com/kubernetes-sigs/metrics-server/issues/572#issuecomment-698183845 still stands, I there are no helm users on Metrics Server team, we don't think we can maintain the quality of chart long term. I have proposed that 2 maintainers to step up, to ensure that the chart is not abandoned. Looking at the fact that before we got second maintainer the first one resigned, I'm not very optimistic if we can fulfill the proposed requirement.

With recent deprecation of kube-state-metrics chart out of official repo, I think we need a larger SIG discussion about supporting Helm. To make sure there is a conclusion to this discussion I would propose to bring this to Regular SIG Instrumentation meeting. https://github.com/kubernetes/community/tree/master/sig-instrumentation#meetings
/cc @logicalhan @ehashman

stevehipwell commented:
@serathius once you have a Helm chart the general maintenance is the same as for your  yaml. That said how would you want to appoint a maintainer for the Helm chart? I'd be happy to offer, there are a couple of other sig charts that I've got an interest in too.

bugok commented:
@serathius: From a user's perspective, having a helm chart reduces the adoption barrier for the service. There are several kubernetes-sigs services which I've decided to use like ([descheduler](https://github.com/kubernetes-sigs/descheduler), [aws load balancer controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller), [external-dns](https://github.com/kubernetes-sigs/external-dns) as well as some non-kubernetes-sigs services ([fluentbit](https://fluentbit.io/), [kube-prometheus-stack](https://github.com/prometheus-operator/kube-prometheus)) and having Helm charts for those made it much easier to install across my multiple clusters. (I'm running everything on top of terraform, so I'm using the Terraform Helm provider for this. Because of this - Having Helm charts makes it very easy to add more services to my clusters).

Today, I'm hard-coding the metrics-server yamls, which isn't great. I'd very much like to have a Helm chart for metrics-server. 

In term of potential usage, I imagine that getting download stats for existing kubernetes-sigs Helm charts could help gauge potential usage of a Helm chart for metrics-server.

krmichel commented:
> @serathius once you have a Helm chart the general maintenance is the same as for your yaml. That said how would you want to appoint a maintainer for the Helm chart? I'd be happy to offer, there are a couple of other sig charts that I've got an interest in too.

I could probably also help maintain the chart.  I don't know that it will have a lot of churn moving forward.

pierluigilenoci commented:
![tenor](https://user-images.githubusercontent.com/36044817/120653623-283f6280-c481-11eb-881d-34063eab6164.gif)

serathius commented:
> That said how would you want to appoint a maintainer for the Helm chart?

@stevehipwell There is no specific process for appointing. Introducing dedicated maintainers will allow to ensure high quality of the chart and give a point of contact in case of any issues. Having two increases long term chance of success, by splitting work and increasing bus factor. I would prefer if maintainers were [Kubernetes members](https://github.com/kubernetes/community/blob/master/community-membership.md#member) to avoid total separation between helm and component contributors, but I don't think it's nesesery. As the PR author and expected main maintainer, feel free to pick a second maintainer, only restriction would be that they are from different company.

@bugok I'm not fully convinced by examples you presented. Maybe I'm doing something wrong but I didn't find any active helm charts belonging to Kubernetes-SIGs org https://artifacthub.io/packages/search?page=1&org=kubernetes-sigs.

>In term of potential usage, I imagine that getting download stats for existing kubernetes-sigs Helm charts could help gauge potential usage of a Helm chart for metrics-server.

@bugok Collecting stats for download is a great idea. Can you take a look into that and provide some numbers?

@krmichel That's great, please coordinate with @stevehipwell 

With two maintainers found and PR updated to reflect that I'm inclined to merge it if we get consensus with other maintainers and encouragement from other SIG Instrumentation TLs/subproject owners.
* @s-urbaniak do you have anything else against introducing support for official Helm chart?
* @brancz or @lilic based on your experience on kube-state-metrics chart, would you recommend Metrics Server to publish official Helm chart?

bugok commented:
@serathius 

> I didn't find any active helm charts belonging to Kubernetes-SIGs org

I meant charts of services belong to Kubernetes-SIGs. I didn't mean that the charts themselves belong to Kubernetes-SIGs. 

About usage stats: Frankly, I don't know how to pull that data in. I didn't see any API which allows to get the stats. However, I did open https://github.com/aws/eks-charts/issues/533 to see if we could get data for one example. If you know of such an API which allows to pull usage stats - I'd be happy to use it.

lilic commented:
> based on your experience on kube-state-metrics chart, would you recommend Metrics Server to publish official Helm chart?

No, we removed ours after initially adding it, despite having two maintainers who use the helm chart, it was a lot of work and it got in the way of the actual main repo releases due to the noise. I would suggest k8s org find a place outside the main repos. Prometheus org does something like this, by adding it to the https://github.com/prometheus-community/helm-charts.

brancz commented:
In the case of kube-state-metrics, it took us multiple attempts of threatening to remove the chart entirely until people really stepped up and are now seriously maintaining it. The reality of open source is, if you want something, then you need to put your time where your mouth is, and not just demand it from maintainers who reap no benefit and only get burden.

I find it a little absurd that in Kubernetes the expectation is that packaging of some arbitrarily chosen mechanism are expected to live alongside the repository, when that's practically nowhere the case for any packages within Linux distributions. I for one would much prefer the same kind of separation in Kubernetes, but I can come to a compromise if there are people who do genuinely commit time to maintaining it.

If this is accepted, I would make it clear on this PR and within the readme that the helm chart may be removed at any point in time should it find itself unmaintained, though personally I would also prefer if helm charts lived outside of application repositories.

stevehipwell commented:
@serathius I'm happy to maintain this with @krmichelos as they've put their hand up to help; do you have any thoughts on adding  a third maintainer from the Kubernetes members?

@brancz I'd personally agree that the ideal location for a chart should be separate but there doesn't seem to be a consensus on this (I'd appreciate you thoughts on the [kubernetes-sig-apps thread](https://groups.google.com/g/kubernetes-sig-apps/c/FB4WgACGQoo/m/4jUp0SwxAgAJ) on this. In the meanwhile having an official chart in the repo seems to be the best option available, but I agree on your caveats around maintenance.

pierluigilenoci commented:
@stevehipwell I tried to upgrade the chart from 3.2.2 to 3.3.0 and I get this error, whether I'm using version 0.4.* or 0.5.0 of metrics-server.

```shell
Error: failed to create listener: failed to listen on 0.0.0.0:443: listen tcp 0.0.0.0:443: bind: permission denied
Usage:
   [flags]

Flags:
      --add_dir_header                                          If true, adds the file directory to the header of the log messages
      --alsologtostderr                                         log to standard error as well as files
      --authentication-kubeconfig string                        kubeconfig file pointing at the 'core' kubernetes server with enough rights to create tokenreviews.authentication.k8s.io.
      --authentication-skip-lookup                              If false, the authentication-kubeconfig will be used to lookup missing authentication configuration from the cluster.
      --authentication-token-webhook-cache-ttl duration         The duration to cache responses from the webhook token authenticator. (default 10s)
      --authentication-tolerate-lookup-failure                  If true, failures to look up missing authentication configuration from the cluster are not considered fatal. Note that this can result in authentication that treats all requests as anonymous.
      --authorization-always-allow-paths strings                A list of HTTP paths to skip during authorization, i.e. these are authorized without contacting the 'core' kubernetes server.
      --authorization-kubeconfig string                         kubeconfig file pointing at the 'core' kubernetes server with enough rights to create subjectaccessreviews.authorization.k8s.io.
      --authorization-webhook-cache-authorized-ttl duration     The duration to cache 'authorized' responses from the webhook authorizer. (default 10s)
      --authorization-webhook-cache-unauthorized-ttl duration   The duration to cache 'unauthorized' responses from the webhook authorizer. (default 10s)
      --bind-address ip                                         The IP address on which to listen for the --secure-port port. The associated interface(s) must be reachable by the rest of the cluster, and by CLI/web clients. If blank or an unspecified address (0.0.0.0 or ::), all interfaces will be used. (default 0.0.0.0)
      --cert-dir string                                         The directory where the TLS certs are located. If --tls-cert-file and --tls-private-key-file are provided, this flag will be ignored. (default "apiserver.local.config/certificates")
      --client-ca-file string                                   If set, any request presenting a client certificate signed by one of the authorities in the client-ca-file is authenticated with an identity corresponding to the CommonName of the client certificate.
      --contention-profiling                                    Enable lock contention profiling, if profiling is enabled
  -h, --help                                                    help for this command
      --http2-max-streams-per-connection int                    The limit that the server gives to clients for the maximum number of streams in an HTTP/2 connection. Zero means to use golang's default.
      --kubeconfig string                                       The path to the kubeconfig used to connect to the Kubernetes API server and the Kubelets (defaults to in-cluster config)
      --kubelet-certificate-authority string                    Path to the CA to use to validate the Kubelet's serving certificates.
      --kubelet-client-certificate string                       Path to a client cert file for TLS.
      --kubelet-client-key string                               Path to a client key file for TLS.
      --kubelet-insecure-tls                                    Do not verify CA of serving certificates presented by Kubelets.  For testing purposes only.
      --kubelet-port int                                        The port to use to connect to Kubelets. (default 10250)
      --kubelet-preferred-address-types strings                 The priority of node address types to use when determining which address to use to connect to a particular node (default [Hostname,InternalDNS,InternalIP,ExternalDNS,ExternalIP])
      --kubelet-use-node-status-port                            Use the port in the node status. Takes precedence over --kubelet-port flag.
      --log-flush-frequency duration                            Maximum number of seconds between log flushes (default 5s)
      --log_backtrace_at traceLocation                          when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                                          If non-empty, write log files in this directory
      --log_file string                                         If non-empty, use this log file
      --log_file_max_size uint                                  Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                                             log to standard error instead of files (default true)
      --metric-resolution duration                              The resolution at which metrics-server will retain metrics. (default 1m0s)
      --permit-port-sharing                                     If true, SO_REUSEPORT will be used when binding the port, which allows more than one instance to bind on the same address and port. [default=false]
      --profiling                                               Enable profiling via web interface host:port/debug/pprof/ (default true)
      --requestheader-allowed-names strings                     List of client certificate common names to allow to provide usernames in headers specified by --requestheader-username-headers. If empty, any client certificate validated by the authorities in --requestheader-client-ca-file is allowed.
      --requestheader-client-ca-file string                     Root certificate bundle to use to verify client certificates on incoming requests before trusting usernames in headers specified by --requestheader-username-headers. WARNING: generally do not depend on authorization being already done for incoming requests.
      --requestheader-extra-headers-prefix strings              List of request header prefixes to inspect. X-Remote-Extra- is suggested. (default [x-remote-extra-])
      --requestheader-group-headers strings                     List of request headers to inspect for groups. X-Remote-Group is suggested. (default [x-remote-group])
      --requestheader-username-headers strings                  List of request headers to inspect for usernames. X-Remote-User is common. (default [x-remote-user])
      --secure-port int                                         The port on which to serve HTTPS with authentication and authorization. If 0, don't serve HTTPS at all. (default 443)
      --skip_headers                                            If true, avoid header prefixes in the log messages
      --skip_log_headers                                        If true, avoid headers when opening log files
      --stderrthreshold severity                                logs at or above this threshold go to stderr (default 2)
      --tls-cert-file string                                    File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated after server cert). If HTTPS serving is enabled, and --tls-cert-file and --tls-private-key-file are not provided, a self-signed certificate and key are generated for the public address and saved to the directory specified by --cert-dir.
      --tls-cipher-suites strings                               Comma-separated list of cipher suites for the server. If omitted, the default Go cipher suites will be used.
                                                                Preferred values: TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA, TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256, TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256, TLS_RSA_WITH_3DES_EDE_CBC_SHA, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_256_GCM_SHA384.
                                                                Insecure values: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_ECDSA_WITH_RC4_128_SHA, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_RSA_WITH_RC4_128_SHA, TLS_RSA_WITH_AES_128_CBC_SHA256, TLS_RSA_WITH_RC4_128_SHA.
      --tls-min-version string                                  Minimum TLS version supported. Possible values: VersionTLS10, VersionTLS11, VersionTLS12, VersionTLS13
      --tls-private-key-file string                             File containing the default x509 private key matching --tls-cert-file.
      --tls-sni-cert-key namedCertKey                           A pair of x509 certificate and private key file paths, optionally suffixed with a list of domain patterns which are fully qualified domain names, possibly with prefixed wildcard segments. The domain patterns also allow IP addresses, but IPs should only be used if the apiserver has visibility to the IP address requested by a client. If no domain patterns are provided, the names of the certificate are extracted. Non-wildcard matches trump over wildcard matches, explicit domain patterns trump over extracted names. For multiple key/certificate pairs, use the --tls-sni-cert-key multiple times. Examples: "example.crt,example.key" or "foo.crt,foo.key:*.foo.com,foo.com". (default [])
  -v, --v Level                                                 number for the log level verbosity
      --version                                                 Show version
      --vmodule moduleSpec                                      comma-separated list of pattern=N settings for file-filtered logging

panic: failed to create listener: failed to listen on 0.0.0.0:443: listen tcp 0.0.0.0:443: bind: permission denied

goroutine 1 [running]:
main.main()
	/go/src/sigs.k8s.io/metrics-server/cmd/metrics-server/metrics-server.go:39 +0xfc
```

Version 3.2.2 of your chart works with both (0.4.* or 0.5.0). I am not able to tell what you changed between 3.2.2 and 3.3.0 but it broke my installation. Could you please take a look?

pierluigilenoci commented:
@stevehipwell found it.
The only changes are the following:

From:
```yaml
- args:
  - --secure-port=4443
[..]
ports:
- containerPort: 4443
  hostPort: 4443
```
To:
```yaml
- args:
  - --secure-port=443
  - --metric-resolution=15s
[..]
ports:
- containerPort: 443
  hostPort: 443
```
We use PSP but, despite the policy, the pod is not able to bind the port. 🤕

pierluigilenoci commented:
@stevehipwell Maybe you can put the port number in the values file in the chart so everyone chooses it as they like. 😘

stevehipwell commented:
@pierluigilenoci what are your custom values? The changes are to align with the official project yaml. I can make the container port customisable if needed.

On a related note I'm not sure we should be supporting the _PodSecurityPolicy_? I only added it as it was in the stable chart and with _PodSecurityPolicy_ now deprecated I've a mind to remove it before this is released.

pierluigilenoci commented:
@stevehipwell PSPs have been deprecated in the old form but from version 1.22 of K8s they will come in a new form. Even if they fail to insert them in version 1.22, the PSPs in the old form will remain at least until version 1.25 so there is time to make educated choices. Ref: https://github.com/kubernetes/enhancements/pull/2582

I would say that it is premature to remove them from the chart also because the SIG said during the [discussion](https://www.youtube.com/watch?v=YSzphARKjKs&t=2340s) that they will most likely be compatible with the old implementation.

My custom values are pretty straightforward:
```yaml
hostNetwork:
  enabled: true
image:
  tag: v0.5.0
nodeSelector:
  node-selector-key: system
rbac:
  pspEnabled: true
resources:
  requests:
    cpu: 10m
    memory: 23.8Mi
service:
  labels:
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: Metrics-server
serviceAccount:
  name: metrics-server
tolerations:
- key: CriticalAddonsOnly
  operator: Exists
```

stevehipwell commented:
@pierluigilenoci have you tested the official YAMl directly?

https://github.com/kubernetes-sigs/metrics-server/releases/download/v0.5.0/components.yamll

pierluigilenoci commented:
> @pierluigilenoci have you tested the official YAMl directly?

@stevehipwell tried with the same result.

stevehipwell commented:
@pierluigilenoci what K8s provider out of interest? Our platform pre-release built on top of EKS got the change this morning and we're seeing the exact same issue as you with psp off.

If you want to open an issue on this repo for the port issue in the official yaml, I'll update the chart to default back to 4443 with it as a variable.

serathius commented on charts/metrics-server/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+sources:
+  - https://github.com/kubernetes-sigs/metrics-server
+maintainers:
+  - name: stevehipwell
```
Please update with a second maintainer.

stevehipwell replied:
@krmichelos are you happy to be the second maintainer?

endrec replied:
If he is not, I'm happy to volunteer. 😃

krmichel replied:
I am fine with it. Please use krmichel as the user name. I am largely afk for about a week.

krmichel replied:
Adding @endrec would give us a third to help out since he said he is willing

stevehipwell replied:
@krmichel my thoughts exactly. Do either of you want an email address or URL added? I'm adding my GitHub profile as my URL.

endrec replied:
You can use my gh profile link too as a URL.

krmichel replied:
My gh profile is fine as well

stevehipwell replied:
Added.

junior replied:
If still time, I'm also willing to help maintain.

stevehipwell replied:
If @serathius is happy for me to add another maintainer and to re-run the tests I'll add you. Otherwise you can always open a PR after this has been merged.

pierluigilenoci replied:
@stevehipwell for me it makes sense to close this as soon as possible and then to add/remove maintainers can always be done later.

stevehipwell commented:
@pierluigilenoci the chart has been updated.

pierluigilenoci commented:
> @pierluigilenoci what K8s provider out of interest? Our platform pre-release built on top of EKS got the change this morning and we're seeing the exact same issue as you with psp off.

@stevehipwell we use this chart on our EKS clusters. So yes, virtually the same infrastructure you have. 

> If you want to open an issue on this repo for the port issue in the official yaml, I'll update the chart to default back to 4443 with it as a variable.

I'll do it.

krmichel commented:
@stevehipwell do we need to be added to the owners file to approve/merge chart changes or do the owners prefer to continue to be the gate check for that?

stevehipwell commented:
@krmichel that's a question for the maintainers I think. @serathius how do you want this to work?

serathius commented:
I would prefer let chart maintainers own the PR review process, but I don't think that's possible if you are not members of  Org. Kubernetes CI bot will not accept lgtm commands from non org members.

endrec commented:
@serathius Can we become org members, pretty please?

Ok, I suppose it is a bit more (but not much more) difficult than that, I believe the biggest obstacle is that we need to find two sponsors.

stevehipwell commented:
@endrec @serathius the sponsors part seems to be the blocker.

s-urbaniak commented:
First, I truly admire the engagement to have an official place of helm charts of metrics-server and the effort put in this PR and discussion :pray: However i personally still believe that the removal of a central helm repository and thus the lack of a central place for helm charts cannot be solved by transferring the burden on single repositories. There needs to be some idiomatic solution, discussed within the helm community how to solve that problem generally. metrics-server is just one tiny piece of the puzzle.

As mentioned above maintainers often lack helm knowledge or have to deal with vanishing sub-maintainership for the helm charts. Additionally, we'd need to make sure these charts stay up-to-date and functional, covered with e2e tests. The cost of maintainership thus is not only limited to text file maintainership.

If based on the discussion in https://groups.google.com/g/kubernetes-sig-apps/c/FB4WgACGQoo/m/4jUp0SwxAgAJ?pli=1 there is no intention to maintain https://github.com/kubernetes-sigs/helm-charts I think we should then at least establish kubernetes-sigs wide rules for helm-charts in general, not only metrics-server, that clearly point out expectations towards the maintainership of those. This would cover expectations around e2e integration and maintainership expectations where repository owners have the right to remove charts if they become unmaintained for some time or if there is lack of quality.

stevehipwell commented:
@s-urbaniak you might be interested in the changes in the last [aws-vpc-cni release](https://github.com/aws/amazon-vpc-cni-k8s/releases/tag/v1.8.0) where they're managing the chart source in the code repo and using CI to push the changes into a central repo to publish the charts. I suspect the reason behind this is that charts and code are tightly coupled and it doesn't work to separate them; obviously if the public contract for the code stays the same the chart needs no changes other than version.

pablitoc commented:
I'm super excited for this! I've been following for a while now and looks like its just about ready. By chance, is there an eta? I'd really like to use this in my new EKS cluster as a pose to installing the metrics server with kubectl.

KlavsKlavsen commented:
BTW. helm chart can be hosted "in the repo" directly.. OR one can use Helm OCI (experiemental since release of 3.0 but works fine) - to host on ghcr.io (github packages repo - ie. docker registry) - that can be enabled for this project.

pierluigilenoci commented:
![source](https://user-images.githubusercontent.com/36044817/128988115-e7a68430-3825-406a-8d98-e89fc557e57f.gif)

https://www.youtube.com/watch?v=2WN0T-Ee3q4

stevehipwell commented:
@serathius I've updated the chart docs and modified the release action to be triggered by a commit to the chart path on the master branch; this means that the release process for the chart is to open a PR to modify _charts/metrics-server/chart.yaml_ with the release version after it has been released (this chart release PR could also contain chart changes if required). Chart changes could also be independently released but this is a choice for the maintainers in if they accept PRs or request that they be merged into the next release PR.

It'd be great to get this merged and I don't see an issue with maintaining this chart without being part of the org, although I'd like to see if I can become a member in due course.

petewilcock commented:
Latest reminder for the maintainers to please, for the love of god, do something with this.

pierluigilenoci commented:
"The saddest PR in history."
Soon on Discovery Channel.
Stay tuned!

![giphy2](https://user-images.githubusercontent.com/36044817/131328740-5728849b-9299-48eb-b4b1-1ff888ccc3fb.gif)

stevehipwell commented:
To show the value of the Helm chart, PR https://github.com/kubernetes-sigs/metrics-server/pull/805 is enabling a workflow that the Helm chart already supports and allows users to customise to their needs (such as more replicas or zone rather than host anti-affinity).

pierluigilenoci commented:
@stevehipwell I really admire your determination.
Thanks for your efforts.

stevehipwell commented:
@pierluigilenoci it's not a huge amount of work to keep the PR tracking correctly, it would be even easier to maintain the chart though, so it's worth doing with the aim of being ready to go. FYI this worked for _External DNS_ (https://github.com/kubernetes-sigs/external-dns/pull/2208).

serathius commented:
Hey, sorry for late response. I've after my vacations I missed some GitHub notifications and was not able to keep up with new ones. Next time please reach out via a Slack https://kubernetes.slack.com/messages/sig-instrumentation it's much reliable way to reach out to maintainers.

I think we are ok with chart 3 maintainers. I would also want to discuss how we can maintain chart quality to be on par with rest of the repo. Topics that I would want to cover:
* How we make sure that future changes to binary don't break chart. Could we setup same e2e tests that we run for release?
* Should default configuration of chart differ from release manifests? Improvements to MS come not only via code, but also default configuration in manifests (for example in v0.5.0 we have increase metric collection rate thanks to performance improvements in code). How such changes be done in chart? 

I would like to also establish exact conditions that both sides  (chart maintainers and repo maintainers) can agree that chart is unmaintained and should be removed. My proposed conditions:
* Chart e2e tests are consistently failing for longer than a month
* Chart is blocking MS release for longer than a month
* There is a issue/PR opened to chart without any response for longer than a quarter
* There is a known bug in chart reported as an issue that is unaddressed for longer than quarter

I hope that those conditions will make it clear what requirements are put on maintainers. I'm also only 1 out of two MS maintainers, would be good to make sure that we have consensus on our side. 

@s-urbaniak do you think that we can introduce an MS chart on conditions proposed above?

endrec commented:
> proposed conditions

I can't speak for @stevehipwell and @krmichel, but those conditions look reasonable to me.

stevehipwell commented:
@serathius sorry I wasn't aware that the Slack channel was there for this.

RE chart configuration; the chart defaults could be the same as for the default manifests, but with the ease of Helm configuration and the fact that values are expected to be set on a chart for real world use the chart might be best implemented with slightly different defaults. I think that before accepting chart changes the MS maintainers would validate that the alignment was as expected, I've been reviewing the changes to the manifests and then doing a diff between the current and previous release manifests to keep the chart aligned.

RE binary compatibility; the chart wont pass CI if the binary is erroring. We could easily add additional Helm tests to the chart for a lightweight E2E check or with more effort add a full E2E suite. However what there currently is should be sufficient considering that the chart release is a secondary process which can be gated and with the configuration guidance above having the chart tracking already tested configuration.

With the way that the chart is released it might be easier to define "unmaintained" as one of the following.

- No MS chart PR to increment app version within a month from the MS release
- Failure to agree on MS chart PR between MS chart maintainers and MS maintainers

The beauty of the current chart design is that a release PR should only need a couple of versions changed in _Chart.yaml_ if the public API hasn't changed. Obviously if best practice for the resources change the chart can change with them. Another benefit is that the chart can make conditional decisions based on context; the current manifests for HA use a `PodDisruptionBudget` API version only supported in K8s v1.21 but the chart can select the correct API for the K8s version (I'm in the process of adding this).

pierluigilenoci commented:
![rogue-one-verge-of-greatness](https://user-images.githubusercontent.com/36044817/131487030-84f7a740-2a61-40e6-a21a-ec45f054c3cb.gif)

s-urbaniak commented:
I am positive on the proposed conditions :+1: @serathius

serathius commented:
/ok-to-test

serathius commented:
/retest

stevehipwell commented:
@serathius could you approve the [lint-test action](https://github.com/kubernetes-sigs/metrics-server/actions/runs/1194180399)?

junior commented:
This PR does not include the `artifacthub-repo.yaml` to publish to [artifacthub.io](artifacthub.io). The are any plans to include it?

stevehipwell commented:
@junior the _artifacthub-repo.yaml_ file is added to the _gh-pages_ branch so can't be part of this PR.

@serathius before this is merged GitHub pages needs to be enabled on this repo with an empty _gh-pages_ branch. If you plan on registering this repo in Artifact Hub then a _artifacthub-repo.yaml_ file should be added to the empty _gh-pages_ branch.

pierluigilenoci commented:
We are so close... what is missing? Just a `/lgtm` label?

serathius commented:
@stevehipwell I have added ghpages. Are there any other prerequisites that are not part of the PR?

As for E2e, I'm not convinced that we should skip them for Helm chart. If we are providing alternative official installation method, we should require same quality. Fact that helm release is independent, doesn't mean that we should not test chart. I think we should run same set of presubmit tests that ensure that both binary and chart changes don't lead to breaking master branch.

This for sure will put maintenance cost on both chart and subproject owners, it will require to introduce and maintain the CI to ensure that PRs don't break compatibility between chart and manifest. However this cost should be smaller than aggregated cost of fixing the release. 

Before merging the change I would want to make sure that we have agreement on the test plan for the chart. I think that aside of chart own tests, we should make sure that chart is also covered by MS e2e test. Those test assume that there is a cluster with MS available and validate that:

* Metrics API is available and gives response any other pod in cluster
* Metrics API exposes metrics about all nodes in cluster
* When deploying CPU consumer with preassigned usage, it returns accurate CPU usage
* When deploying Memory consumer with preassigned usage, it returns accurate Memory usage
* Properly handles sidecar and init containers
* Passes health and readiness probes
* Exposes expected metrics

code: https://github.com/kubernetes-sigs/metrics-server/blob/master/test/e2e_test.go

All those tests are independent on MS deployment method and should pass on both manifests and chart. I would want to add running e2e tests for charts to run as part of PR, but not be required to pass. This way we can detect early breakage and assign chart owners to fix it.

stevehipwell commented:
@serathius as long as the `gh-pages` branch isn't protected what you've done should be good. You might also want to [configure Artifact Hub](https://artifacthub.io/docs/topics/repositories/#helm-charts-repositories) both for discovery and to show that the chart is verified.

RE E2E testing that makes sense but could we get the chart merged first so the configuration of this can be added in a separate PR by someone who knows how the MS E2E setup is configured better than I do? That said I'd be happy to review the PR for this to make sure that the chart was being installed with the correct values.

serathius commented:
> RE E2E testing that makes sense but could we get the chart merged first so the configuration of this can be added in a separate PR by someone who knows how the MS E2E setup is configured better than I do? That said I'd be happy to review the PR for this to make sure that the chart was being installed with the correct values.

Sounds Good,
/lgtm

Leaving approval to @s-urbaniak

s-urbaniak commented:
Thank you everybody! :bow: 
/approve

onedr0p commented:
Sitting from the top of the roof, a big thanks to everyone involved.

stevehipwell commented:
@serathius @s-urbaniak I'll create a PR to fix the release workflow which seems to be broken.

samhopwell commented:
This is amazing news, thank you everyone!

stevehipwell commented:
@serathius the release workflow is functioning correctly but the `gh-pages` branch is currently protected which [isn't supported](https://github.com/kubernetes-sigs/metrics-server/pull/670#issuecomment-913536138), the _cluster-autoscaler_ team, who use the same workflow, started off with a protected `gh-pages` branch but have now unprotected it. The solution is to un-protect the branch for future releases and then generate and manually add the _index.yaml_ file to `gh-pages`. I could open a PR to `gh-pages` with the index in it?

serathius commented:
sg, will unprotect the branch.

serathius commented:
done

stevehipwell commented:
@serathius I've created https://github.com/kubernetes-sigs/metrics-server/pull/822 to publish the chart.

stevehipwell commented:
Thanks to the hard work by @serathius we now have the official Metrics Server helm chart published.

```shell
helm repo add metrics-server https://kubernetes-sigs.github.io/metrics-server/
helm repo update
helm search repo metrics-server
```

Now we just need to get the repo registered in Artifact Hub.

pierluigilenoci commented:
I can't believe it! ❤️ 
Thank you all!!!

![giphy](https://user-images.githubusercontent.com/36044817/133077366-3d33cd16-60e7-4308-93f5-25fd759fc752.gif)
